_posts/2020-09-07-introduction-to-system-tests.md

```diff
@@ -0,0 +1,258 @@
+                        .endKafkaListenerExternalNodePort()
+                    .endListeners()
+                .endKafka()
+            .endSpec()
+            .build();
+    KafkaCluster kc = KafkaCluster.fromCrd(kafkaAssembly, VERSIONS);
+
+    // Check StatefulSet changes
+    StatefulSet sts = kc.generateStatefulSet(true, null, null);
+
+    List<ContainerPort> ports = sts.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts();
+    assertThat(ports.contains(kc.createContainerPort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, "TCP")), is(true));
+
+    // Check external bootstrap service
+    Service ext = kc.generateExternalBootstrapService();
+    assertThat(ext.getMetadata().getName(), is(KafkaCluster.externalBootstrapServiceName(cluster)));
+    assertThat(ext.getSpec().getType(), is("NodePort"));
+    assertThat(ext.getSpec().getSelector(), is(kc.getSelectorLabels().toMap()));
+    assertThat(ext.getSpec().getPorts(), is(Collections.singletonList(kc.createServicePort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, KafkaCluster.EXTERNAL_PORT, "TCP"))));
+    checkOwnerReference(kc.createOwnerReference(), ext);
+
+    // Check per pod services
+    for (int i = 0; i < replicas; i++)  {
+        Service srv = kc.generateExternalService(i);
+        assertThat(srv.getMetadata().getName(), is(KafkaCluster.externalServiceName(cluster, i)));
+        assertThat(srv.getSpec().getType(), is("NodePort"));
+        assertThat(srv.getSpec().getSelector().get(Labels.KUBERNETES_STATEFULSET_POD_LABEL), is(KafkaCluster.kafkaPodName(cluster, i)));
+        assertThat(srv.getSpec().getPorts(), is(Collections.singletonList(kc.createServicePort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, KafkaCluster.EXTERNAL_PORT, "TCP"))));
+        checkOwnerReference(kc.createOwnerReference(), srv);
+    }
+}
+```
+
+The next one is the integration tests, which step one level higher.
+You can find all these tests inside our `api` module.
+Before we get into detail again we should understand the informal definition of integration tests.
+
+> Integration tests determine if independently developed units of software work correctly when they are connected to each other. (Martin Fowler)
+
+In our case the purpose of integration tests is to ensure that we can create a resource from the POJOs, serialize it and
+create the resource in K8S.
+You should notice that in that level of testing we are using the Kubernetes cluster to do validation of custom resources.
+To make this all easy to understand let me clarify it with example:
+
+```java
+@Test
+public void testKafkaWithEntityOperator() {
+    createDelete(Kafka.class, "Kafka-with-entity-operator.yaml");
+}
+
+protected <T extends CustomResource> void createDelete(Class<T> resourceClass, String resource) {
+    T model = loadResource(resourceClass, resource);  <--- load model in that case Kafka model representation
```

Comment:
Wouldn't be better to put your comment part as regular Java comment above the line?